### PR TITLE
Adds ActorRef.ignore

### PR DESCRIPTION
**Pull request by octonato**
References #25306

### CODE REVIEW

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+package akka.cluster.typed
+
+import akka.actor.typed.{ ActorRef, ActorRefResolver, ActorSystem }
+import akka.actor.typed.scaladsl.Behaviors
+import com.typesafe.config.ConfigFactory
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class ActorRefIgnoreSerializationSpec extends AnyWordSpec with ScalaFutures with Matchers {
```
Comment (renatocaval) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: For the serialization, I choose to add this extra test on the cluster typed module. Otherwise, I would have to add cluster deps in test scope in akka-actor-typed-tests.

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      val system1 = ActorSystem(Behaviors.empty[String], "sys1", config)
+      val system2 = ActorSystem(Behaviors.empty[String], "sys2", config)
+
+      try {
+
+        val remoteRefStr = ActorRefResolver(system1).toSerializationFormat(ActorRef.ignore[String])
+        remoteRefStr should (startWith("akka://sys1@127.0.0.1") and endWith("/system/ignore"))
+
+        val ref: ActorRef[String] = ActorRefResolver(system2).resolveActorRef[String](remoteRefStr)
+        ref.path.toString should (startWith("akka://sys1@127.0.0.1") and endWith("/system/ignore"))
```
Comment (renatocaval) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: I would prefer to have the path set to `sys2`. As it is now, the reply will be sent back to system1 and will be ignored there. It will be better if we can simply ignore it locally.

Comment (renatocaval) [reply akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: I managed to solve this. See second commit.

File: akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala [hunk 3]
```diff
@@ -524,6 +524,13 @@ private[akka] class RemoteActorRefProvider(
    * public `resolveActorRef(path: String)`.
    */
   private[akka] def internalResolveActorRef(path: String): ActorRef = path match {
+
+    case p if IgnoreActorRef.isIgnoreRefPath(p) =>
+      // we need a local address to avoid sending messages that will be ignored over the wire
+      // basically, we pretend that this Ref belongs to this system
+      val localAddressToUse = transport.localAddressForRemote(systemGuardian.path.address)
+      new IgnoreActorRef(localAddressToUse.hostPort)
```
Comment (renatocaval) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: This is the trick that guarantees that a IgnoreActorRef that is deserialized on another node, won't send unnecessary messages over the wire. We pretend it's a local Actor.

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 4]
```diff
@@ -476,6 +476,33 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] object IgnoreActorRef {
+
+  val pathElements: Iterable[String] = immutable.Seq("system", "ignore")
+  val pathString = pathElements.mkString("/", "/", "")
```
Comment (renatocaval) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: Hmm, I will change the name of this two vals.

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 5]
```diff
@@ -476,6 +476,33 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+
+  def isIgnoreRefPath(path: String): Boolean = path.endsWith(pathString)
```
Comment (renatocaval) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: This check is fragile. We need to check teh queality for the whoe=le part after the host and port. I will change this.

File: akka-docs/src/main/paradox/typed/interaction-patterns.md [hunk 6]
```diff
@@ -225,6 +225,38 @@ Java
...
  * When `ask` times out, the receiving actor does not know and may still process it to completion, or even start processing it after the fact
 
+## Ignoring replies
+
+In some situations you may need to send a message that requires a @scala[`replyTo:ActorRef[T]`]@java[`ActorRef<T> replyTo`], but you are not interested in receiving the reply. In this case, you can pass a @scala[`ActorRef.ignore`]@java[`ActorRef.ignore()`].
+
+@scala[`ActorRef.ignore`]@java[`ActorRef.ignore()`], as the name indicates, returns an `ActorRef` that ignores any message sent to it.
+
+
+With the given protocol and actor behavior:
```
Comment (johanandren) [inline akka-docs/src/main/paradox/typed/interaction-patterns.md]: What about re-using the request-response sample and referring to that rather than a full protocol just to show of passing ignore? 

"With the same protocol as the @ref[request response](...) above, ignoring the reply would look like this:

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+      // this is more a compile-time proof
+      // since the reply is ignored, we can't check that a message was sent to it
+      pingPongRef ! Ping(ActorRef.ignore)
+
+      // but we do check that the counter has increased when we sent the ActorRef.ignore
+      pingPongRef ! Ping(probe.ref)
+      probe.expectMessage(Pong(3))
+    }
+  }
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: Some unittytesty things perhaps - will two `ignore` compare the same?

I think we should also cover some more corner cases here, if nothing else just to document expected behavior:

What if the recipient (unknowingly that it is ignore):
 * tries to ask the ref something (response-with-ack)
 * tries to watch the ref
 * adapt to classic and use that as sender in interaction with classic actor - which leads to
   * future.pipeTo(ignore) 
   * classic forward(ignore)
   * classic router.route
 * any more I'm not coming up with from the top of my head?

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+      try {
+
+        val remoteRefStr = ActorRefResolver(system1).toSerializationFormat(ActorRef.ignore[String])
+        // the remote ref string path must be the actor sys name with host and port
+        // concatenated with IgnoreActorRef relative path
+        remoteRefStr shouldBe IgnoreActorRef.actorPath(system1.address.hostPort).toString
+
+        val ref: ActorRef[String] = ActorRefResolver(system2).resolveActorRef[String](remoteRefStr)
+        ref.path.toString shouldBe IgnoreActorRef.actorPath(system2.address.hostPort).toString
```
Comment (johanandren) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: When it is a single or actor-system single instance, also compare instance.

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java [hunk 9]
```diff
@@ -896,6 +938,23 @@ public void timers() throws Exception {
...
+
+    // note that system is also the ActorRef to the guardian actor
+    final ActorRef<PingPong.Ping> ref = system;
+
+    // sends a message using 'tell' and ignores the reply, fire-and-forget style
+    ref.tell(new PingPong.Ping(ActorRef.ignore()));
+    // #ignore-reply-doit
+
+    system.terminate();
+    system.getWhenTerminated().toCompletableFuture().get(5, TimeUnit.SECONDS);
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java]: Can use `ActorTestKit.shutdown(...); instead. That will also await the termination. I don't know why other places here don't use it. Good if you change all.

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala (hunk 8 above)
Comment (renatocaval) [reply akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: I don't get your point here.

Comment (johanandren) [reply akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: If you deserialize ignore from another system it should be the same instance of ignore as in this system, not create a new object every time - feedback above, so should be verified here when that is the case.

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRef.scala [hunk 10]
```diff
@@ -134,3 +140,13 @@ object RecipientRef {
...
+
+/**
+ * INTERNAL API
+ *
+ * An ActorRef that ignores any incoming messages.
+ * This reference cannot watch other references.
+ *
+ */
+@InternalApi private[typed] final class IgnoreActorRef[T]()
+    extends ActorRefAdapter[T](new classic.IgnoreActorRef("local"))
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/ActorRef.scala]: I wonder if the class is needed, or if it would work with a:
```
private val ignoreRef = ActorRefAdapter[Any](new classic.IgnoreActorRef("local"))
```
and 
```
def ignore[T]: ActorRef[T] = ignoreRef.unsafeCast[T] // safe cast
```

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 11]
```diff
@@ -476,6 +476,47 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+ * An ActorRef that ignores any incoming messages.
+ *
+ * INTERNAL API
+ */
+private[akka] class IgnoreActorRef(systemName: String) extends MinimalActorRef {
+
+  override def provider: ActorRefProvider =
+    throw new UnsupportedOperationException("no provider")
+
+  override def path: ActorPath = IgnoreActorRef.actorPath(systemName)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: `val` ?

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 12]
```diff
@@ -476,6 +476,47 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: Might also need `writeReplace` to support Java serialization. See how that is done in for example `akka.actor.Nobody.

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 13]
```diff
@@ -476,6 +476,47 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+  /**
+   * Check if the passed `path` has the same relative path of the IgnoreActorRef (ie: /system/ignore).
+   * @param path
+   * @return
+   */
+  def isIgnoreRefPath(path: String): Boolean = {
+    val pathElements = ActorPath.fromString(path).toStringWithoutAddress
+    // whatever sysName is good, we ony want to take the relative path
+    val ignorePathElements = actorPath("any-system").toStringWithoutAddress
+    pathElements == ignorePathElements
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: To make this more efficient we don't have to make it an ActorPath but can use some string matching.

File: akka-docs/src/main/paradox/typed/interaction-patterns.md [hunk 14]
```diff
@@ -225,6 +225,38 @@ Java
...
+
+Scala
+:  @@snip [InteractionPatternsSpec.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala) { #ignore-reply-doit }
+
+Java
+:  @@snip [InteractionPatternsTest.java](/akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java) { #ignore-reply-doit }
+
+**Useful when:**
+
+  * Sending a message requiring a `replyTo`, but you are not interested on the reply
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/interaction-patterns.md]: interested in

File: akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala [hunk 15]
```diff
@@ -524,6 +524,13 @@ private[akka] class RemoteActorRefProvider(
    * public `resolveActorRef(path: String)`.
    */
   private[akka] def internalResolveActorRef(path: String): ActorRef = path match {
+
+    case p if IgnoreActorRef.isIgnoreRefPath(p) =>
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: good thing is that those are cached before calling internalResolveActorRef

**Review by patriknw — COMMENTED** [review]
looking good, great work @renatocaval

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 16]
```diff
@@ -476,6 +476,47 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+ */
+private[akka] object IgnoreActorRef {
+
+  /**
+   * Returns the ignore actor path for the passed actor system name.
+   * @param systemName
+   * @return
+   */
+  def actorPath(systemName: String): ActorPath =
+    RootActorPath(Address("akka", systemName)) / "system" / "ignore"
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: on the other hand that name collision can happen by other usages of systemActorOf too

[the remark this answers is not in the document]
Comment (johanandren) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: Yeah, but this is more like `/deadLetters` than the extension system actors though.

[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: true, maybe there is no drawback to create it at the top level?

**Comment by johanandren** [comment]
One more thing I thought about while background processing this is how it is dealt with in the artery actor ref cache - it could perhaps always be cached?

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala (hunk 8 above)
Comment (renatocaval) [reply akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: This is not possible because of the trick I did [here](https://github.com/akka/akka/pull/28630/files/9490cc8e6eb8d42b983ba91c429ef5878ef43eda#r382329107) to avoid sending messages over the wire, but also because `ActorRef.ignore` doesn't depend on an ActorSystem.

However, we can prove that: 
 * `ActorRef.ignore[String] isSameAs ActorRef.ignore[String]`
 * `deserRef isSameAs otherDeserRef` becasue the deserialized instance is cached.

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 17]
```diff
@@ -476,6 +476,47 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
 
+/**
+ * An ActorRef that ignores any incoming messages.
+ *
+ * INTERNAL API
+ */
+private[akka] class IgnoreActorRef(systemName: String) extends MinimalActorRef {
+
+  override def provider: ActorRefProvider =
+    throw new UnsupportedOperationException("no provider")
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: good catch, there is no way around that and it's actually used in many places

 have to be system.ignore then, like deadLetters but silent

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala (hunk 7 above)
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: Tests for ask and watch was added. Also conversion to classic back and forth. 

As discussed, we are postponing the decision to add more tests to cover integration with Akka classic.

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala [hunk 18]
```diff
@@ -82,11 +82,19 @@ class InteractionPatternsSpec extends ScalaTestWithActorTestKit with AnyWordSpec
...
       // #request-response-send
 
       probe.receiveMessage()
+
+      Behaviors.setup { context =>
+        // #ignore-reply
+        cookieFabric ! CookieFabric.Request("don't send cookies back", context.system.ignoreRef)
+        // #ignore-reply
+        Behaviors.empty[Nothing]
+      }
```
Comment (renatocaval) [inline akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala]: I'm not adding any new test. No probe since any message is ignore anyway and this has already been tested elsewhere.

File: akka-docs/src/main/paradox/typed/interaction-patterns.md [hunk 19]
```diff
@@ -76,7 +76,7 @@ Java
 :  @@snip [InteractionPatternsTest.java](/akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java) { #request-response-protocol }
 
 
-The sender would use its own @scala[`ActorRef[Response]`]@java[`ActorRef<Response>`], which it can access through @scala[`ActorContext.self`]@java[`ActorContext.getSelf()`], for the `respondTo`. 
+The sender would use its own @scala[`ActorRef[Response]`]@java[`ActorRef<Response>`], which it can access through @scala[`ActorContext.self`]@java[`ActorContext.getSelf()`], for the `replyTo`.
```
Comment (renatocaval) [inline akka-docs/src/main/paradox/typed/interaction-patterns.md]: the example in the snippet uses `replyTo`

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala [hunk 20]
```diff
@@ -56,6 +55,7 @@ abstract class ActorRefResolver extends Extension {
       ref.path.toSerializationFormatWithAddress(system.address)
 
     ref.toClassic match {
+      case _: IgnoreActorRef => IgnoreActorRef.path.toString // IgnoreActorRef never changes its path
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala]: this will not be handled by classic serialization of such ref

Alt 1)
Special treatment in toSerializationFormatWithAddress, but that is a bit difficult because there we don't easily know that it's an ignoreRef.

Alt 2)
Don't fake the serializationFormat but just use ordinary toSerializationFormatWithAddress? Then when deserializing  we must figure out that it's a ignoreRef.

I'm not sure which alternative has less performance impact on serialization of ordinary refs, which is important.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+  def behavior(askMeRef: ActorRef[Request], probe: TestProbe[Int]) = Behaviors.setup[Int] { context =>
+    implicit val timeout: Timeout = 1.second
+
+    // send a message to interactWithRef
+    context.ask(askMeRef, Request) {
+      case Success(res) => res
+      case Failure(ex)  => throw ex
+    }
+
+    def internalBehavior(): Behavior[Int] = {
```
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: That's a leftover I guess. Will remove.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    // this is kind of obvious, the Future won't complete because the ignoreRef is used
+    "make a Future timeout when used in a 'ask'" in {
+
+      implicit val timeout: Timeout = 500.millis
+      implicit val scheduler = testKit.system.scheduler
+      val askMeRef = testKit.spawn(askMeActorBehavior)
+
+      eventually {
```
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: The `eventually` is in the wrong place here. It should only wrap the last line (`failure shouldBe a[TimeoutException]`). 

That future should timeout because we pass the ignoreRef to the command

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 23]
```diff
@@ -368,16 +368,26 @@ final class ChildActorPath private[akka] (val parent: ActorPath, val name: Strin
...
-    appendUidFragment(sb).toString
+    if (IgnoreActorRef.isIgnoreRefPath(this)) {
+      // we never change address for IgnoreActorRef
+      this.toString
+    } else {
+      val diff = addressStringLengthDiff(addr)
+      val length = toStringLength + diff
+      val sb = buildToString(new JStringBuilder(length + 12), length, diff, _.toStringWithAddress(addr))
+      appendUidFragment(sb).toString
+    }
```
Comment (renatocaval) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: @patriknw, is that what you meant? In `ActorPath.toSerializationFormatWithAddress`?

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 24]
```diff
@@ -368,16 +368,26 @@ final class ChildActorPath private[akka] (val parent: ActorPath, val name: Strin
...
-    val length = toStringLength + diff
-    buildToString(new JStringBuilder(length), length, diff, _.toStringWithAddress(addr)).toString
+    if (IgnoreActorRef.isIgnoreRefPath(this)) {
+      // we never change address for IgnoreActorRef
+      this.toString
+    } else {
+      val diff = addressStringLengthDiff(addr)
+      val length = toStringLength + diff
+      buildToString(new JStringBuilder(length), length, diff, _.toStringWithAddress(addr)).toString
+    }
```
Comment (renatocaval) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: I changed this one just to align with `toSerializationFormatWithAddress`

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala [hunk 25]
```diff
@@ -76,7 +76,10 @@ abstract class ActorRefResolver extends Extension {
   }
 
   override def resolveActorRef[T](serializedActorRef: String): ActorRef[T] =
-    classicSystem.provider.resolveActorRef(serializedActorRef)
+    if (IgnoreActorRef.isIgnoreRefPath(serializedActorRef))
+      classicSystem.toTyped.ignoreRef
+    else
+      classicSystem.provider.resolveActorRef(serializedActorRef)
 }
```
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala]: I understand and agree, but the problem here is that if we move the impl to `resolveActorRef` only classic ref will be cached. 

`classicSystem.toTyped.ignoreRef` always returns the same instance that is already wrapped on an adapter. Otherwise, we will have a new adapter for each invocation.

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+    }
+  }
+
+  "IgnoreActorRef (in classic)" should {
+
+    "return a serializable ActorRef that can be sent between two ActorSystems using remote (akka classic)" in {
+
+      val ignoreRef = system1.ignoreRef[String].toClassic
+      val remoteRefStr = ignoreRef.path.toSerializationFormatWithAddress(system1.address)
```
Comment (renatocaval) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorRefIgnoreSerializationSpec.scala]: Is this how I should generate a remoteRef String in classic? I didn't find any thing beter. Please, let me know if this can be improved.

**Review by octonato — COMMENTED** [review]
@patriknw, did some more changes after your last review. And left some comments as well.

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala (hunk 25 above)
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala]: btw, `classic.provider.resolveActorRef` is also taking care of this. We have code duplication here to avoid the adapter instance.

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala [hunk 27]
```diff
@@ -141,6 +141,15 @@ abstract class ActorSystem[-T] extends ActorRef[T] with Extensions with ClassicA
    */
   def deadLetters[U]: ActorRef[U]
 
+  // this needs to be a lazy val because some impl of ActorSystem
+  // (eg: ActorSystemStub) does not have a provider
+  private lazy val cachedIgnoreRef: ActorRef[Nothing] = ActorRefAdapter(provider.ignoreRef)
```
Comment (renatocaval) [inline akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala]: Builds were surprisingly failing all over the place because of ActorSystemStub. It's unfortunate that we need to make it lazy because of test concerns.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+        askMeRef
+          .ask { _: ActorRef[Request] =>
+            Request(testKit.system.ignoreRef) // <- pass the ignoreRef instead, so Future never completes
+          }
+          .failed
+          .futureValue(PatienceTimeout(1.second))
+
+      eventually {
+        failedAsk shouldBe a[TimeoutException]
+      }
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: `eventually` means that it will retry the block, but here there is nothing to retry. The Future has already been completed, and waited for with `futureValue`. I think this works fine without eventually.

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala (hunk 27 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala]: And it would be unfortunate if ignoreRef can't be used with ActorSystemStub (synchronous behavior testing).
I think it would be better to override the `def ignoreRef` implementation in `ActorSystemStub` and return a DebugRef there, in the same way as `deadLetters`.

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala]: regarding the lazy, this can be moved to `ActorSystemAdapter` and leave `def ignoreRef` unimplemented (abstract) in ActorSystem

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 29]
```diff
@@ -476,6 +476,51 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+  private val fakeSystemName = "local"
+
+  val path: ActorPath =
+    RootActorPath(Address("akka", IgnoreActorRef.fakeSystemName)) / "ignore"
+
+  /**
+   * Check if the passed `otherPath` is the same as IgnoreActorRef.path
+   */
+  def isIgnoreRefPath(otherPath: String): Boolean =
+    path.toString == otherPath
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: to avoid constructing new strings each time we can have
```
private val pathString = path.toString
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala (hunk 25 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala]: maybe slightly better performance (for ordinary refs) with this:
```scala
override def resolveActorRef[T](serializedActorRef: String): ActorRef[T] = {
  val ref = classicSystem.provider.resolveActorRef(serializedActorRef)
  if (ref eq classicSystem.ignoreRef)
    classicSystem.toTyped.ignoreRef // use cached ActorRefAdapter
  else
    ref
}
```

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 30]
```diff
@@ -476,6 +476,51 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
...
+
+  /**
+   * Check if the passed `otherPath` is the same as IgnoreActorRef.path
+   */
+  def isIgnoreRefPath(otherPath: ActorPath): Boolean =
+    path == otherPath
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: good

**Review by patriknw — COMMENTED** [review]
a few comments, but this is shaping up

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 31]
```diff
@@ -476,6 +476,51 @@ private[akka] trait MinimalActorRef extends InternalActorRef with LocalRef {
   protected def writeReplace(): AnyRef = SerializedActorRef(this)
 }
 
+/**
+ * An ActorRef that ignores any incoming messages.
+ *
+ * INTERNAL API
+ */
+@InternalApi private[akka] class IgnoreActorRef(override val provider: ActorRefProvider) extends MinimalActorRef {
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: `final`?

**Review by johanandren — COMMENTED** [review]
Looking good, added a few small notes

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala [hunk 32]
```diff
@@ -82,11 +82,19 @@ class InteractionPatternsSpec extends ScalaTestWithActorTestKit with AnyWordSpec
       val context = new {
         def self = probe.ref
```
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala]: I will see I I can remove this hack and still having something to test the documentation sample, but not a priority.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala (hunk 28 above)
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: correct, I will remove it

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala (hunk 32 above)
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala]: Yeah, not important to do anything about IMO

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 33]
```diff
@@ -368,16 +368,26 @@ final class ChildActorPath private[akka] (val parent: ActorPath, val name: Strin
   }
 
   override def toStringWithAddress(addr: Address): String = {
-    val diff = addressStringLengthDiff(addr)
-    val length = toStringLength + diff
-    buildToString(new JStringBuilder(length), length, diff, _.toStringWithAddress(addr)).toString
+    if (IgnoreActorRef.isIgnoreRefPath(this)) {
```
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor/src/main/scala/akka/actor/ActorPath.scala]: Is that really worth? 

It will mean that every single ActorPath will do this comparison at constructor time while now it's only done when we need it. 

Unless we make it a lazy val

[the remark this answers is not in the document]
Comment (johanandren) [reply akka-actor/src/main/scala/akka/actor/ActorPath.scala]: Most ActorPaths probably do not see any calls to either of these two methods, but I think paths that do are more likely to see several calls (called from serialization since  they are used in a remoting context). 

Lazy would be bad because that would be a synchronization on each access to the boolean, since it is the outcome of a calculation that never changes it could be lazy-calculated without synchronization/volatile but since when serializing there are likely a lot of other more expensive things going on so let's leave as is for now.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala (hunk 7 above)
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: Any other tests ideas or we can consider this one as resolved for now?

Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorRefIgnoreSpec.scala]: Looks good enough for me

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala (hunk 32 above)
[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-actor-typed-tests/src/test/scala/docs/akka/typed/InteractionPatternsSpec.scala]: Well, I had a look at it. The code is lying. It pretends to send a `Response` to `context.self`, but `Response` is not part of `CookieFabric`'s protocol. And it compiles because it's the probe, not the real self. 

I will send another PR to fix this.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]
LGTM
